core/src/main/scala/core/database/QueryRunner.scala

```diff
@@ -4,19 +4,20 @@ import scala.concurrent.{ExecutionContext, Future}
 
 import play.api.db.Database
 
+import core.utils.Id
+
 class QueryRunner[DB <: Database](
-    db: Database,
-    ec: ExecutionContext
+  db: Database,
+  ec: ExecutionContext
 ) {
-  def run[A](query: Query[A]): Future[A] =
-    Future {
-      db.withConnection(query.f)
-    }(ec)
-
-  // TODO: Remove with a transaction run at context level
-  def commit[A](query: Query[A]): Future[A] =
+  def run[A, C](query: Query.Aux[A, C])(implicit ev: Id[C]): Future[A] = // TODO: maybe rename it to commit ?
```

Comment:
There I don't see why it would not be possible to consider the contextual runner as `Query[T] => Future[A]` (or even as `Reader[Connection, A] => Future[A]`).

```scala
trait QueryRunner {
  def apply[T](query: Query[T]): Future[T]
  // an runner instance be used following: runner(query)
}

final class SimpleRunner {
  def apply[T](query: Query[T]) = db.withConnection(query.underlying.f)
}

final TransactionalRunner {
  def apply[T](query: Query[T]) = db.withTransaction(query.underlying.f)
}
```

As stream `Materializer` (which represents strategy to materialize runnable graph), it could be propagated implicitly, in which case following could be useful.

```scala
object QueryRunner {
  def run[T](query: Query[T])(implicit runner: QueryRunner): Future[T] = runner(query)
}
```

Reply:
Yes, that was what I figured when re-thinking about this, but here we need to disambiguate between the two `implicit`s (with a tag on `Connection` or an evidence). How could we choose the right running strategy if we don't keep a track of the context in the `Query`?

Reply:
I don't see the need to track. The query must be totally agnotic about the context(es) with which it will then be eventually executed.

> Tracking = coupling with context